core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -122,33 +132,41 @@ public void setPartitionCount(int partitionCount)
             if (collected) {
                 return;
             }
-            checkState(expectedPartitionCount == PARTITION_COUNT_INITIAL_VALUE, "setPartitionCount should be called only once");
-            requireNonNull(partitions, "partitions is null");
+            checkState(expectedPartitionCount == null, "setPartitionCount should be called only once");
             expectedPartitionCount = partitionCount;
-            if (partitions.size() == expectedPartitionCount) {
-                // No more partitions are left to be processed.
-                if (partitions.isEmpty()) {
-                    result = TupleDomain.none();
-                }
-                else {
-                    result = TupleDomain.columnWiseUnion(partitions);
-                }
-                collected = true;
-                partitions = null;
+            if (collectedPartitionCount < expectedPartitionCount) {
+                return;
+            }
+            if (partitionCount == 0) {
+                result = TupleDomain.all();
             }
             else {
-                return;
+                // run final compaction as previous concurrent compactions may have left more than a single domain
+                compactSummaryDomains();
+                verify(summaryDomains.size() == 1);
+                result = summaryDomains.poll();
+                verify(result != null);
             }
+            collected = true;
         }
 
-        notifyConsumers(result);
+        collectors.forEach(collector -> collector.accept(convertTupleDomain(result)));
     }
 
-    private void notifyConsumers(TupleDomain<DynamicFilterId> result)
+    private void compactSummaryDomains()
     {
-        requireNonNull(result, "result is null");
-        Map<DynamicFilterId, Domain> dynamicFilterDomains = convertTupleDomain(result);
-        collectors.forEach(consumer -> consumer.accept(dynamicFilterDomains));
+        while (true) {
+            TupleDomain<DynamicFilterId> first = summaryDomains.poll();
+            if (first == null) {
+                return;
+            }
+            TupleDomain<DynamicFilterId> second = summaryDomains.poll();
+            if (second == null) {
+                summaryDomains.add(first);
+                return;
+            }
+            summaryDomains.add(columnWiseUnion(first, second));
```

Comment:
nit: It might be simpler to read if written as
```
ImmutableList.Builder<TupleDomain<DynamicFilterId>> builder = ImmutableList.builder();
TupleDomain<DynamicFilterId> head = summaryDomains.poll();
while (head != null) {
     builder.add(head);
     head = summaryDomains.poll();
}
List<TupleDomain<DynamicFilterId>> tupleDomains = builder.build();
if (!tupleDomains.isEmpty()) {
      summaryDomains.add(columnWiseUnion(tupleDomains));
}
```

Reply:
It will cost extra allocations that seem to be unnecessary (in happy path there will be only two domains to merge). But the cost of extra allocations shouldn't be very significant, so I don't have a very strong opinion here.

Reply:
I am fine with the code but worth a comment that common case is that only up to two domains are in the queue